ngraph_bridge/ngraph_data_cache.h

```diff
@@ -0,0 +1,180 @@
+#include "tensorflow/core/graph/graph.h"
+
+#include "logging/ngraph_log.h"
+#include "ngraph/ngraph.hpp"
+
+#include "ngraph_bridge/ngraph_freshness_tracker.h"
+#include "ngraph_bridge/ngraph_pipelined_tensors.h"
+
+namespace tensorflow {
+
+namespace ngraph_bridge {
+
+// Forward declaration for friend class
+namespace testing {
+class NGraphDataCacheTest_SameKeyMultiThread_Test;
+class NGraphDataCacheTest_RemoveItemTest_Test;
+}
+
+template <typename KeyType, typename ValueType>
+class NgraphDataCache {
+ public:
+  explicit NgraphDataCache(int depth);
+  ~NgraphDataCache();
+
+  // This method performs lookup in the cache for requested key, if not found
+  // it will create item, put it in the cache and returns item and status
+  std::pair<Status, ValueType> LookUpOrCreate(
+      KeyType key,
+      std::function<std::pair<Status, ValueType>()> callback_create_item,
+      std::function<void(ValueType)> callback_destroy_item);
+
+  // Overload for above function, which doesn't take callback for destroy item
+  std::pair<Status, ValueType> LookUpOrCreate(
+      KeyType key,
+      std::function<std::pair<Status, ValueType>()> callback_create_item);
+  void RemoveItem(KeyType key);
+  void RemoveItem(KeyType key,
+                  std::function<void(ValueType)> callback_destroy_item);
+  void RemoveAll(std::function<void(ValueType)> callback_destroy_item);
+
+ private:
+  std::unordered_map<KeyType, ValueType> m_ng_items_map;
+  std::deque<KeyType> m_lru;
+  int m_depth;
+  absl::Mutex m_mutex;
+
+  // Test class
+  friend class tensorflow::ngraph_bridge::testing::
+      NGraphDataCacheTest_SameKeyMultiThread_Test;
+  friend class tensorflow::ngraph_bridge::testing::
+      NGraphDataCacheTest_RemoveItemTest_Test;
+};
+
+template <typename KeyType, typename ValueType>
+NgraphDataCache<KeyType, ValueType>::NgraphDataCache(int depth)
+    : m_depth(depth) {}
+
+template <typename KeyType, typename ValueType>
+NgraphDataCache<KeyType, ValueType>::~NgraphDataCache() {
+  m_ng_items_map.clear();
+}
+
+template <typename KeyType, typename ValueType>
+void NgraphDataCache<KeyType, ValueType>::RemoveItem(
+    KeyType key, std::function<void(ValueType)> callback_destroy_item) {
+  absl::MutexLock lock(&m_mutex);
+  if (m_ng_items_map.find(key) != m_ng_items_map.end()) {
+    callback_destroy_item(m_ng_items_map.at(key));
+    m_ng_items_map.erase(key);
+    m_lru.pop_back();
+  }
+}
+
+template <typename KeyType, typename ValueType>
+void NgraphDataCache<KeyType, ValueType>::RemoveItem(KeyType key) {
+  RemoveItem(key, [](ValueType) {});
+}
+
+template <typename KeyType, typename ValueType>
+void NgraphDataCache<KeyType, ValueType>::RemoveAll(
+    std::function<void(ValueType)> callback_destroy_item) {
+  absl::MutexLock lock(&m_mutex);
+  for (auto it = m_ng_items_map.begin(); it != m_ng_items_map.end(); it++) {
+    callback_destroy_item(it->second);
+  }
+  m_ng_items_map.erase(m_ng_items_map.begin(), m_ng_items_map.end());
```

Comment:
erase m_lru? all keys from m_lru should also be deleted?

Reply:
Done